Imporve/fix some scopes reported by the JavaScript parser

b4n opened the pull request.

masatake commented:
For newly added test cases, could you set --sort=no option?
Just put "--sort=no" to args.ctags file.

(In the future, I will put --sort=no to misc/units script.)

b4n commented:
@masatake OK, but for JS I don't know if it makes much sense as the parser emits some tags out of order.

b4n commented:
@masatake actually, what's the point in not sorting?  I agree it's easier to check when the parser is emitting things in the order they appear in the file, but otherwise I kind of feel like the tests would relying too much on an implementation detail that never mattered.  Namely, I don't have a really sensible way of "not sorting" the new test I added and that doesn't pass yet: which order should I put?  None is really relevant.
Again, I'm not sure why you don't want sorting, but maybe we should have a mean to sort by line numbers?

masatake commented:
@b4n,  how about adding `--fields=+n` (or `--fields=+{line}`) instead of `--sort`?

masatake commented:
I will review them in the week end. Thank you. BTW, is this change something to do with
https://github.com/universal-ctags/ctags/issues/1566#issuecomment-334462281 ?

b4n commented:
@masatake thanks!  I just squashed the fixups and added `--fields=+n` to the failing test case as you suggested.
And no, it's unrelated to that issue, yet related to Geany.  I was in the process of updating Geany's parser to current U-CTags for https://github.com/geany/geany/issues/934 and realized one of the test cases was giving bad results, so I went ahead and spend a day fixing it -- which itself lead to finding other small issues and trying to fix them, and trying to increase the coverage.

masatake commented on Units/parser-javascript.r/js-scope-resolution.b/expected.tags:
```diff
@@ -0,0 +1,10 @@
+dummy1	input.js	/^    this.dummy1 = function() {}$/;"	m	line:18	class:doesnt_extend.Cls
+dummy2	input.js	/^  Cls.prototype.dummy2 = function() {$/;"	m	line:20	class:doesnt_extend.Cls
+extend	input.js	/^function extend() {$/;"	f	line:7
+extended	input.js	/^  Cls.prototype.extended = function() {$/;"	m	line:9	class:Cls
```
Very interesting.

"Cls" in the scope field is obviously useful information about "extended".
However, "extend" in which "extended" is assigned is also useful information.

None defines the meaning of "scope" field. I have just recognized that there are two type of scopes: 
source code aspect scope and runtime scope.  lexical scope and dynamic scope are more suitable terms?

For reading and understanding source code, source code aspect scope information is also useful. The 
information can be regenerated with line: and end: fiedls but recording it in ctags is not bad idea.

Some parsers may records runtime scope and the other parsers may records source code aspect scope as the value for scope field. It may be too late to redefine the meaning of scope field. However, I would like to explore this area in the future. 

@b4n, you have developed an IDE. I would like to hear your comments.

One of my idea is introducing new scope named sourceScope that is filled only when if source code aspect scope and runtime scope about a tag is not matched.

masatake commented on Units/parser-javascript.r/js-crlf.d/expected.tags:
```diff
@@ -0,0 +1,12 @@
+v5	input.js	/^,   v5 = 3$/;"	v
+f1_c1_m1	input.js	/^    f1_c1_m1:$/;"	m	class:f1.f1_c1
+f1_c1_m2	input.js	/^    f1_c1_m2$/;"	m	class:f1.f1_c1
+f1_c1	input.js	/^  var f1_c1 = {$/;"	c	class:f1
```
It seems that f1 is a function.

b4n replied:
Indeed. As mentioned somewhere it's a known limitation of the JS parser, with a `FIXME` in the code.

masatake commented:
Have you ever consider to use corkAPI for representing scopes in jscritp.c?

For filling a scope field, a parser have to track both scope name and scope kind.
tokenInfo of jscript.c has only a structure field for scope name. 
Adding a field of scope kind is an approach.
Using corkAPI is another approach.

In my experince, corkAPI works well in this case.

I can take over the rest of job for introducing corkAPI if you are o.k. In such case, I will merge this pull request first.

masatake commented:
@b4n, after rebasing the lastet master branch where CTAGS_INLINE macro is introduced, could you consider chery-picking following changes?

https://github.com/masatake/ctags/commit/690c456fd9f160fe2ac6248a8b912dde5cd82f63
https://github.com/masatake/ctags/commit/d4a27f0461e11605a68ff937e6b23c33c7ca4f07
https://github.com/masatake/ctags/commit/cacb33fc3e428bba37c24f2e505b294e67d6689c
https://github.com/masatake/ctags/commit/524dd1bd298d7385b8631cb866bfc976b52fbce2

All of them are in found at https://github.com/masatake/ctags/commits/jscript-make-signature-field-sophisticated

b4n commented on Units/parser-javascript.r/js-commas-and-missing-semicolons.d/expected.tags:
```diff
@@ -0,0 +1,11 @@
+b	input.js	/^var A = {}, b = 2;$/;"	v
+c	input.js	/^var c = 3, d = 4;$/;"	v
+d	input.js	/^var c = 3, d = 4;$/;"	v
+mem1	input.js	/^A.mem1 = function() {return 42;}, A.mem2 = function() {return 43;}$/;"	f	function:A
```
The problem is that in JavaScript there's not really anything but objects, a class, a function, no real differences.  Actually you can see a fun thing in Firefox's JS console:
```
>> class foo {}
undefined
>> print(foo)
"function foo() {
    
}"
```

Anyway, my point is that the effective high-level kind is very dynamic.  In this example, A should be a class merely because it gets methods added, but it really becomes a class.  So ideally A should be reported as a class, but when it's first seen and the tag emitted it looks like a naked object, hence the "variable" kind.

b4n commented:
@masatake I though about using that new cork API, but it would require quite some changes, especially as many tags are held back (to find the right kind).  Maybe though it would be quite handy if we can change the kind of a tag after having prepared it.
Anyway, I'm not against using it at all, I just felt it would be a fairly larger undertaking I didn't want to start right away, but I'd be fine reviewing the changes, or trying it at some point.

b4n commented:
> I think introducing "unknown" kind is better.

Hum… maybe, I'm not sure.  OK, it doesn't report something wrong and a client *could* special-case it, but OTOH the parser is generally right in the assumption, only a few cases are incorrect, so I'm not sure.

masatake commented:
I found we have js-scope-resolution.b test case.
I'm sorry to be late to know that. 

LGTM. Thank you.

masatake commented:
@b4h, should I merge this?

b4n commented:
@masatake You can, yes.
I am in the process of importing your mentioned changes, but wanted to try and simplify a few things, but didn't get there yet.  It's however mostly unrelated to this PR, so I think merging would be fine.
And I have another set of changes for the JS parser (for supporting Unicode identifiers and character escapes) in the works, so it's not like this one would be the last one anyway ☺️

masatake commented:
Could you look at#1519? It will change the last argument of initTagEntry. It required a pointer to kindDefintion struct. With the change of #1519, it requires an integer instead. This change may have small impact on the code you are working.
